lib/config/config.js

```diff
@@ -111,24 +110,50 @@ class Config {
     return this._updateConfigFile('system', modifier);
   }
 
-  // this function can preserve comments and formatting.
-  // due to fragile nature of YAWN it may not work properly in that mode.
-  // safe way is to use it in rewrite file mode
   _updateConfigFile (type, modifier) {
-    const path = this[this.configTypes[type].pathProperty];
+    const configType = this.configTypes[type];
+    const path = this[configType.pathProperty];
+
     return readFile(path, 'utf8').then(data => {
-      let text;
-      const isJSON = path.toLowerCase().endsWith('.json');
-      if (this.systemConfig && this.systemConfig.preserveStructureOnUpdates && !isJSON) {
-        const yawn = new YAWN(data);
-        yawn.json = modifier(yawn.json);
-        text = yawn.yaml;
-      } else { // js-yaml can handle JSON as well as YAML if no need for structure\comments save
-        const json = yamlOrJson.load(data);
-        const result = modifier(json);
-        text = yamlOrJson.dump(result);
+      const json = yamlOrJson.load(data);
+      const result = modifier(json);
+      const text = yamlOrJson.dump(result);
+      const candidateConfiguration = yamlOrJson.load(envReplace(String(text), process.env));
+
+      const { isValid, error } = configType.validator(candidateConfiguration);
+
+      if (!isValid) {
+        const e = new Error(error);
+        e.code = 'INVALID_CONFIG';
+        throw e;
+      }
+
+      try {
+        /*
+          This is really bad. It means we have circular dependencies that's a code smell, no matter what. This needs
+          to be refactored as soon as possible.
+        */
+        const { policies } = require('../policies');
+        for (const pipelineName in candidateConfiguration.pipelines) {
+          const pipeline = candidateConfiguration.pipelines[pipelineName];
+
+          pipeline.policies.forEach(policy => {
+            const policyName = Object.keys(policy)[0];
+            const policyDefinition = policies[policyName];
+            const { isValid, error } = schemas.validate(policyDefinition.schema.$id, policy[policyName].action || {});
+            if (!isValid) {
+              throw new Error(error);
+            }
+          });
+        }
+
+        return writeFile(path, text);
+      } catch (err) {
+        log.error('Invalid pipelines configuration.', err);
+        err.code = 'INVALID_CONFIG';
```

Comment:
Should we reuse this logic in `loadConfig`?  In our create-gateway templates, we can also look for this error and display a warning instead of crashing the process with an uncaught promise rejection.

Reply:
In theory I'd like to make the opposite — this part should be using the `loadConfig`. Unfortunately this is not possible in a timely manner because of the coupled code between load config and config source — which is probably going to be a different story.

I can try to reword this piece a little bit, but it might take a little bit of time before having something good; we can create a tracking card for that. Your call.

P.S: I'm using https://github.com/ExpressGateway/express-gateway/projects to track things I'd like to change in Express Gateway for the better which might result in a breaking change. Feel free to have a look into!

Reply:
Okay, let's move forward in the interest of time.  Thanks!